File: languages/csharp/03-advanced/constructors-and-invariants/example/bank-account-guardrails.cs
// This helper example focuses on showing how helper methods preserve invariants around deposits and withdrawals.

// This extra example extends constructors and invariants with a bank account.

using System;

// Keep this helper separate so the main example can focus on the larger idea without extra noise.
class BankAccount
{
    private readonly string owner;
    private double balance;

    public BankAccount(string ownerValue, double initialBalance)
    {
        owner = ownerValue.Trim().Length == 0 ? "Unknown" : ownerValue.Trim();
        balance = initialBalance < 0.0 ? 0.0 : initialBalance;
    }

    public bool Deposit(double amount)
    {
        if (amount <= 0.0)
        {
            return false;
        }

        balance += amount;
        return true;
    }

    public bool Withdraw(double amount)
    {
        if (amount <= 0.0 || amount > balance)
        {
            return false;
        }

        balance -= amount;
        return true;
    }

    public string Summary => $"{owner}: {balance:F2}";
}

class Program
{
    static void Main()
    {
        BankAccount account = new BankAccount("  Ana  ", -50.0);
        Console.WriteLine($"Initial: {account.Summary}");
        Console.WriteLine($"Deposit 20 success: {account.Deposit(20.0)}");
        Console.WriteLine($"Withdraw 10 success: {account.Withdraw(10.0)}");
        Console.WriteLine($"Withdraw 99 success: {account.Withdraw(99.0)}");
        Console.WriteLine($"Final: {account.Summary}");
    }
}

File: languages/python/03-advanced/constructors-and-invariants/example/bank-account-guardrails.py
# This helper example focuses on showing how helper methods preserve invariants around deposits and
# withdrawals.

# This extra example extends constructors and invariants with a bank account.

# Keep this helper separate so the main example can focus on the larger idea without extra noise.
class BankAccount:
    def __init__(self, owner, initial_balance):
        self._owner = owner.strip() or "Unknown"
        self._balance = max(0.0, initial_balance)

    def deposit(self, amount):
        if amount <= 0.0:
            return False

        self._balance += amount
        return True

    def withdraw(self, amount):
        if amount <= 0.0 or amount > self._balance:
            return False

        self._balance -= amount
        return True

    @property
    def summary(self):
        return f"{self._owner}: {self._balance:.2f}"


account = BankAccount("  Ana  ", -50.0)
print(f"Initial: {account.summary}")
print(f"Deposit 20 success: {account.deposit(20.0)}")
print(f"Withdraw 10 success: {account.withdraw(10.0)}")
print(f"Withdraw 99 success: {account.withdraw(99.0)}")
print(f"Final: {account.summary}")
